controllers/machine_controller.go

```diff
@@ -645,6 +666,13 @@ func (r *MachineReconciler) watchClusterNodes(ctx context.Context, cluster *clus
 		Watcher:      r.controller,
 		Kind:         &corev1.Node{},
 		EventHandler: handler.EnqueueRequestsFromMapFunc(r.nodeToMachine),
+		Indexes: []remote.Index{
+			{
+				Obj:          &corev1.Node{},
+				Field:        NodeProviderIDIndex,
+				ExtractValue: indexNodeByProviderID,
+			},
+		},
```

Comment:
I don't think this should be defined in here, but rather we should expose some options to `NewClusterCacheTracker` to add indexes when the ClusterTracker is initialized. 

Only indexes defined at the top level will be able to be used throughout the lifecycle of a ClusterCacheTracker, and if a user tries to access something an unknown index, it's up to them to define them in the list of indexes.

Reply:
that sounds good to me, though it would mean all clusterAccessor/clients for a tracker need to use the same indexes defeating the original attempt indexing only each client for its watched kind https://github.com/kubernetes-sigs/cluster-api/pull/4521#discussion_r620642643
Unless we store some relation kind/index in `NewClusterCacheTracker` and then we use that value to discriminate and pass only the right indexes through `getClusterAccessorLH` from func `Watch`

Reply:
@vincepri updated PTAL.

Reply:
Why should we make the assumption that this client is always going to be used to index nodes?

Reply:
Seemed convenient to me given I don't think of `remote` as a general purpose library but rather as something we are the main consumers and we know we need to index nodes. I'll update to drop the default and set the index when instantiating the tracker.

Reply:
Updated.